File: backends/dotnet/src/Zappy.Domain/Promotions/PromotionCode.cs
namespace Zappy.Domain;

public sealed class PromotionCode
{
    private PromotionCode()
    {
    }

    public PromotionCode(
        string code,
        PromotionKind kind,
        int? percentage,
        Money? amount,
        Money? minimumSubtotal,
        DateTimeOffset validFrom,
        DateTimeOffset validUntil,
        int? usageLimit,
        int timesUsed)
    {
        Code = code.Trim().ToUpperInvariant();
        Kind = kind;
        Percentage = percentage;
        Amount = amount;
        MinimumSubtotal = minimumSubtotal;
        ValidFrom = validFrom;
        ValidUntil = validUntil;
        UsageLimit = usageLimit;
        TimesUsed = timesUsed;
    }

    public string Code { get; private set; } = null!;

    public PromotionKind Kind { get; private set; }

    public int? Percentage { get; private set; }

    public Money? Amount { get; private set; }

    public Money? MinimumSubtotal { get; private set; }

    public DateTimeOffset ValidFrom { get; private set; }

    public DateTimeOffset ValidUntil { get; private set; }

    public int? UsageLimit { get; private set; }

    public int TimesUsed { get; private set; }

    public PromotionRule Rule => Kind switch
    {
        PromotionKind.Percentage => new PercentageOffSubtotal(Percentage ?? 0),
        PromotionKind.FixedAmount => new FixedAmountOffSubtotal(Amount ?? Money.Euro(0)),
        PromotionKind.FreeShipping => new FreeShipping(),
        _ => throw new InvalidOperationException($"{Code} carries the unknown kind {Kind}.")
    };

    public static string Normalise(string code) => code.Trim().ToUpperInvariant();

    public Result<PromotionRule> RuleFor(Money subtotal, DateTimeOffset moment)
    {
        if (moment < ValidFrom || moment > ValidUntil)
        {
            return Result<PromotionRule>.Failure(
                UserErrorCode.CodeExpired,
                $"The promotion code {Code} is outside its validity window.",
                "code");
        }

        if (UsageLimit is not null && TimesUsed >= UsageLimit)
        {
            return Result<PromotionRule>.Failure(
                UserErrorCode.CodeExhausted,
                $"The promotion code {Code} has reached its usage limit.",
                "code");
        }

        if (MinimumSubtotal is not null && !subtotal.IsAtLeast(MinimumSubtotal))
        {
            return Result<PromotionRule>.Failure(
                UserErrorCode.CodeMinimumNotMet,
                $"The promotion code {Code} asks for a subtotal of at least {MinimumSubtotal}.",
                "code");
        }

        return Result<PromotionRule>.Success(Rule);
    }

    public void RecordUse() => TimesUsed += 1;
}

File: backends/node/subgraphs/promotions/src/domain/promotionCode.ts
import type { Money } from "@zappy/shared";
import { isAtLeast, money } from "@zappy/shared";
import type { PromotionKind, PromotionRule } from "./promotionRule.js";
import { fixedAmountRule, freeShippingRule, percentageRule } from "./promotionRule.js";

export type PromotionCode = {
  readonly code: string;
  readonly kind: PromotionKind;
  readonly percentage: number | null;
  readonly amount: Money | null;
  readonly minimumSubtotal: Money | null;
  readonly validFrom: string;
  readonly validUntil: string;
  readonly usageLimit: number | null;
  readonly timesUsed: number;
};

export type CodeVerdict = "usable" | "CODE_EXPIRED" | "CODE_EXHAUSTED" | "CODE_MINIMUM_NOT_MET";

export function normaliseCode(typed: string): string {
  return typed.trim().toUpperCase();
}

export function ruleOf(code: PromotionCode): PromotionRule {
  if (code.kind === "PERCENTAGE") {
    return percentageRule(code.percentage ?? 0);
  }
  if (code.kind === "FIXED_AMOUNT") {
    return fixedAmountRule(code.amount ?? money(0));
  }
  return freeShippingRule;
}

export function judgeCode(code: PromotionCode, subtotal: Money, moment: Date): CodeVerdict {
  const from = new Date(code.validFrom).getTime();
  const until = new Date(code.validUntil).getTime();
  const now = moment.getTime();
  if (now < from || now > until) {
    return "CODE_EXPIRED";
  }
  if (code.usageLimit !== null && code.timesUsed >= code.usageLimit) {
    return "CODE_EXHAUSTED";
  }
  if (code.minimumSubtotal !== null && !isAtLeast(subtotal, code.minimumSubtotal)) {
    return "CODE_MINIMUM_NOT_MET";
  }
  return "usable";
}
